cmd/tracee-ebpf/capabilities.go

```diff
@@ -73,17 +75,66 @@ func getCapabilitiesRequiredByTraceeEvents(cfg *tracee.Config) []cap.Value {
 	return removeDupCaps(caps)
 }
 
+// Retrieve the value of the kernel parameter perf_event_paranoid
+func getKernelPerfEventParanoidValue() int {
+	/*
+	* perf event paranoia level:
+	*  -1 - not paranoid at all
+	*   0 - disallow raw tracepoint access for unpriv
+	*   1 - disallow cpu events for unpriv
+	*   2 - disallow kernel profiling for unpriv
+	*   4 - disallow all unpriv perf event use
+	 */
+	const MAX_PARANOIA_LEVEL = 4
+	value, err := os.ReadFile("/proc/sys/kernel/perf_event_paranoid")
+	if err != nil {
+		return MAX_PARANOIA_LEVEL
+	}
+	intVal, err := strconv.ParseInt(strings.TrimSuffix(string(value), "\n"), 0, 16)
+	if err != nil {
+		return MAX_PARANOIA_LEVEL
+	}
+	return int(intVal)
+}
+
 // Get all capabilities required for eBPF usage (including perf buffers maps management)
-func getCapabilitiesRequiredByEBPF(OSInfo KernelVersionInfo) ([]cap.Value, error) {
+func getCapabilitiesRequiredByEBPF(OSInfo KernelVersionInfo, debug bool) ([]cap.Value, error) {
 	// In kernel 5.8, CAP_BPF and CAP_PERFMON capabilities were introduced in order to replace CAP_SYS_ADMIN when
 	// loading eBPF programs.
-	// For some reasons, some distributions using new kernels still need CAP_SYS_ADMIN,
-	// so tracee still use it instead of the new capabilities.
-	caps := []cap.Value{
+	kernelParanoidValue := getKernelPerfEventParanoidValue()
+	var caps []cap.Value
+	privilegedCaps := []cap.Value{
 		cap.IPC_LOCK,
 		cap.SYS_RESOURCE,
 		cap.SYS_ADMIN,
 	}
+	unprivilegedCaps := []cap.Value{
+		cap.IPC_LOCK,
+		cap.SYS_RESOURCE,
+		cap.BPF,
+		cap.PERFMON,
+	}
+	const bpfCapabilitiesMinKernelVersion = "5.8"
+	if OSInfo.CompareOSBaseKernelRelease(bpfCapabilitiesMinKernelVersion) > 0 {
+		caps = privilegedCaps
+	} else {
+		// if kernelParanoidValue is too high, CAP_SYS_ADMIN is required
+		if kernelParanoidValue > 3 {
+			if debug {
+				fmt.Print("Kernel perf_event_paranoid value is greater than 3, ")
+				fmt.Print("requiring Tracee to use the CAP_SYS_ADMIN capability instead of CAP_BPF + CAP_PERFMON\n")
+				fmt.Println("To remove this message, set the kernel parameter perf_event_paranoid to 3 or less.")
```

Comment:
So I wonder if this is actually a bug (in Ubuntu?) since the kernel documentation says that CAP_PERFMON should be enough for every operation used with perf: https://www.kernel.org/doc/html/latest/admin-guide/perf-security.html#perf-events-access-control 
Actually, the kernel sources don't even mention levels 3 or 4 as can be seen in the comment here:
https://elixir.bootlin.com/linux/v5.19/source/kernel/events/core.c#L416
So if I understand it right, it's probably an addition of Ubuntu, which is not part of mainline kernel, and they "forgot" to allow the operation with CAP_PERFMON enabled, am I right?

Reply:
They extracted from grsecurity patches:

![image](https://user-images.githubusercontent.com/7395852/182519170-ba045dbf-b7e6-4d25-b645-ceba9d7ede5e.png)

Reply:
I think we should assume that there is a deviation in between upstream acceptance of >=2 and grscurity/ubuntu (and possibly other distros ?) having > 2 values.

Reply:
> They extracted from grsecurity patches:
> 
> ![image](https://user-images.githubusercontent.com/7395852/182519170-ba045dbf-b7e6-4d25-b645-ceba9d7ede5e.png)

From when is this commit dated? Maybe they required CAP_SYS_ADMIN before CAP_PERFMON was available (5.8), and simply forgot to add the option for this capability sinve these kernels

Reply:
I think you might be right, as in this discussion (https://patchwork.kernel.org/project/linux-hardening/patch/1469630746-32279-1-git-send-email-jeffv@google.com/) they seem to wish for a `CAP_PERF` that would handle this kind of things.

Reply:
> I think we should assume that there is a deviation in between upstream acceptance of >=2 and grscurity/ubuntu (and possibly other distros ?) having > 2 values.

That really looks like it as I noticed that Tracee runs fine on a recompiled kernel with a paranoid value > 3 without `CAP_SYS_ADMIN`

Reply:
Yep and I can talk to their kernel team and ask them to address that, and will, but we will have to work in many already releases broken kernels even if they fix that in latest releases I'm afraid.

Reply:
Yeah, so I don't really think it changes something to the way this PR addresses the `perf_event_paranoid` value problem